packages/codecs-core/src/decoder-exact-byte-array.ts

```diff
@@ -0,0 +1,45 @@
+import { SOLANA_ERROR__CODECS__BYTES_LENGTH_DECODER_MISMATCH, SolanaError } from '@solana/errors';
+
+import { Decoder } from './codec';
+import { transformDecoder } from './transform-codec';
+
+/**
+ * Create a {@link Decoder} that asserts that the bytes provided to `decode` or `read` are exactly the length needed by the inner decoder
+ * @param decoder A decoder to wrap
+ * @returns A new decoder that will throw if provided with a byte array of the wrong length to decode
+ *
+ * @typeParam T - The type of the decoder
+ *
+ * @remarks
+ * Note that this compares the offset after encoding to the length of the input byte array
+ *
+ * The `offset` parameter to `decode` and `read` is still considered, and will affect the new offset that is compared to the byte array length
+ *
+ * The error that is thrown by the returned decoder is a {@link SolanaError} with the code `SOLANA_ERROR__CODECS__BYTES_LENGTH_DECODER_MISMATCH`
+ *
+ * @example
+ * Create a decoder that decodes a `u32` (4 bytes) and ensures the entire byte array is consumed
+ * ```ts
+ * const decoder = createDecoderThatUsesExactByteArray(getU32Decoder());
+ * decoder.decode(new Uint8Array([0, 0, 0, 0])); // 0
+ * decoder.decode(new Uint8Array([0, 0, 0, 0, 0])); // throws
+ * decoder.decode(new Uint8Array([0, 0, 0])); // throws
+ *
+ * // with an offset
+ * decoder.decode(new Uint8Array([0, 0, 0, 0, 0]), 1); // 0
+ * decoder.decode(new Uint8Array([0, 0, 0, 0, 0, 0]), 1); // throws
+ * decoder.decode(new Uint8Array([0, 0, 0, 0]), 1); // throws
+ * ```
+ */
+export function createDecoderThatUsesExactByteArray<T>(decoder: Decoder<T>): Decoder<T> {
+    return transformDecoder(decoder, (value, bytes, offset, newOffset) => {
+        if (newOffset !== bytes.length) {
+            throw new SolanaError(SOLANA_ERROR__CODECS__BYTES_LENGTH_DECODER_MISMATCH, {
+                bytesLength: bytes.length,
+                offsetAfterDecoding: newOffset,
+                offsetBeforeDecoding: offset,
```

Comment:
This makes people do math. Just `expectedBytes`/`actualBytes` or just `numExcessBytes`?